Add Prettier script and pre-commit check for HTML and stylesheets

nospame opened the pull request:
## Product Description
Dev only. Adds a check to pre-commit to identify if modified HTML or stylesheets could use reformatting, e.g.:

```console
$ git commit -m "billing statements"
corehq/apps/domain/templates/domain/billing_statements.html
HTML and/or stylesheets have code style issues.
Run ./scripts/prettify-changed.sh to fix them in a separate commit.

There are formatting errors in your commit. Commit with --no-verify to proceed without fixing.
```
If there are also Python formatting issues raised by flake8, they will appear first, followed by the above notes about html or stylesheets. The `--no-verify` flag will still work as it currently does, by skipping the check entirely.

```console
$ ./scripts/prettify-changed.sh
Formatting files changed on this branch:
corehq/apps/domain/templates/domain/billing_statements.html 33ms

Remember to commit these changes.
```

## Technical Summary
Adds prettier and prettier-plugin-jinja-template to dev dependencies along with relatively simple shell scripts to use the dependency. As written will require devs to run `yarn install` or they will see an error from node on trying to commit html/css/scss/less files (if using the pre-commit hook).

Prettier was chosen here over other formatters for its speed, multi-format compatibility, ease of use, and minimal dependencies. The downside is it is not exceptionally configurable. If devs find it truly does not format to our liking, we might consider a different (more explicitly configurable) formatter.

## Safety Assurance

### Safety story
This adds two dev dependencies, both widely-used and maintained. Other changes are only in dev scripts, so would have no effect elsewhere.

### Automated test coverage
Nope.

### QA Plan
Not formally, but requesting that devs try it out.


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Add Prettier dependency

nospame committed:
Add Prettier plugin for jinja-style templates

nospame committed:
Add prettierrc with .html-specific config

nospame committed:
Add prettify-staged.sh script

nospame committed:
Add Prettier check to pre-commit

mkangia commented:
Nice!

Can we include in PR description what formatting rules we are adding?
I see we have set tab spaces. or there are no rules and that is just to set the tabsize.

mkangia commented:
fyi @millerdev for the dependencies being added.

millerdev commented on scripts/prettify-staged.sh:
```diff
@@ -0,0 +1,20 @@
+prettify() {
+    FILES=$1
+    echo "Formatting staged and uncommitted files:"
+    ./node_modules/.bin/prettier --write $FILES
```
> staged and uncommitted

Does this include files that are not staged (or only partially staged)? If yes, I would be annoyed if I unsuspectingly ran a script and it modified files that had unstaged changes since it could be difficult to untangle hand-crafted changes from automated prettifier changes.

orangejenny replied:
Yeah, and as a reviewer I much prefer when formatting changes are separate from logical changes. One way to mitigate this is to change the instructions to encourage committing the formatting separately. Another is to reformat the entire repo once this is merged (which I think we should do anyway, it'll take forever to reformat opportunistically).

nospame replied:
In local testing this correctly excludes files that are _not_ staged, but it does include files that are partially staged and formats them as a whole. If that's a major issue, we could use something like [git-format-staged](https://github.com/hallettj/git-format-staged) to get formatting for only those portions of files that are staged.

nospame replied:
@orangejenny would preferable behavior in your mind be something more like this?
- Identify which files want reformatting in pre-commit, but don't block the commit for it
- Recommend running the command to reformat those whole files
- Use the prettify-staged (or a renamed/modified version) script to optionally commit those files as well, similar to the `build_bootstrap5_diffs` command

orangejenny replied:
`git-format-staged` seems nice for its purpose (I'm also a devotee of `-p` when staging changes). I think the downside is that we'll then have files that have old formatting interspersed with new, but that's fixable by then formatting the entire file in a new commit that reviewers can largely ignore.

orangejenny replied:
(commented before seeing your comment addressed to me, and seeing your previous comment gave me more to think about) I think blocking the commit is fine, since `--no-verify` exists. It's good to have an attention-grabbing moment, and blocking the commit does that. I was basically asking to append "in a separate commit" to "Run ./scripts/prettify-staged.sh to fix them automatically."

nospame replied:
@orangejenny That seems reasonable. I think there would also need to be a corresponding change to the behavior of `prettify-staged` for usability. If we're asking a developer to make style fixes in a separate commit, perhaps that script should target all files that have been changed and _are_ committed relative to `master`, and maybe call it `prettify-changed` or something?

nospame commented:
> Nice!
> 
> Can we include in PR description what formatting rules we are adding? I see we have set tab spaces. or there are no rules and that is just to set the tabsize.

@mkangia Yeah, Prettier in particular is really minimal in the options it allows. Aside from [tab width](https://prettier.io/docs/en/options.html#tabs) which we want as 2-space instead of the default 4-space they use, the default options seemed the closest to what we liked. Other relevant options that this currently uses the default for are [print width](https://prettier.io/docs/en/options.html#single-attribute-per-line) (which is different from a "max width"), [bracket line](https://prettier.io/docs/en/options.html#bracket-line), [whitespace sensitivity](https://prettier.io/docs/en/options.html#html-whitespace-sensitivity), and [single attribute per line](https://prettier.io/docs/en/options.html#single-attribute-per-line).

jingcheng16 commented:
@nospame Reviewed but have no questions! Nice work!

nospame committed:
prettify-staged => prettify-changed

Look for already committed files instead of uncommitted, and improve error handling

millerdev commented on git-hooks/pre-commit.sh:
```diff
@@ -54,7 +54,18 @@ if [ $? != 0 ]; then status=1; fi
+    if [ $? != 0 ]; then
+        status=1;
+	    echo "HTML and/or stylesheets above have code style issues."
+	    echo "Run ./scripts/prettify-changed.sh to fix them in a separate commit."
```
This wording seems a little ambiguous about when to run the command. What do you think of
```suggestion
	    echo "After committing changes with --no-verify, run ./scripts/prettify-changed.sh to fix them."
```

millerdev commented on scripts/prettify-changed.sh:
```diff
@@ -0,0 +1,18 @@
+}
+
+CHANGED_FILES=$(get_changed) || { echo "No changed files found to format." && exit 1; }
+prettify "$CHANGED_FILES" || { echo "Oops! Did you remember to run yarn install?" && exit 1; }
```
Is it possible/expected that there will be errors that cause this "Oops! ..." output, but to which it is not a useful suggestion?

nospame replied:
Technically possible if Prettier hits a parse error while reading any of the files passed in, but since we're limiting by filetype this is usually okay -- the rare exception being knockout/embeddedJS type templates that use `<% %>` and do so outside of `<script>` tags. In those cases you'd see something like this from Prettier:
```console
[error] corehq/apps/cloudcare/templates/cloudcare/partials/pagination.html: SyntaxError: Opening tag "a" not terminated. (46:21)
```

I added that "Oops" message because the "no such file or directory" error when `./node_modules/.bin/prettier` isn't found is not particularly helpful.

nospame committed:
Clarify when to run prettify-changed

Co-authored-by: Daniel Miller <dmiller@dimagi.com>
